libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -274,6 +342,53 @@ void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel_
     }
 }
 
+/*
+ * Compute the back away velocity required to avoid breaching margin
+ */
+void AC_Avoid::calc_backaway_velocity(float kP, float accel_cmss, Vector2f &quad1_back_vel_cms, Vector2f &quad2_back_vel_cms, Vector2f &quad3_back_vel_cms, Vector2f &quad4_back_vel_cms, float back_distance_cm, Vector2f limit_direction, float dt)
+{      
+    if (limit_direction.is_zero()) {
+        // protect against divide by zero
+        return; 
+    }
+    // speed required to move away the exact distance that we have breached the margin with 
+    const float back_speed = get_max_speed(kP, 0.4f * accel_cmss, fabsf(back_distance_cm), dt);
+    
+    // direction to the obstacle
+    limit_direction.normalize();
+
+    // move in the opposite direction with the required speed
+    Vector2f back_direction_vel = limit_direction * (-back_speed);
+    // divide the vector into quadrants, find maximum velocity component in each quadrant 
+    find_max_quadrant_velocity(back_direction_vel, quad1_back_vel_cms, quad2_back_vel_cms, quad3_back_vel_cms, quad4_back_vel_cms);
+}
+
+/*
+ *  Calculate maximum velocity vector that can be formed in each quadrant 
+*/
+void AC_Avoid::find_max_quadrant_velocity(Vector2f &desired_vel, Vector2f &quad1_vel, Vector2f &quad2_vel, Vector2f &quad3_vel, Vector2f &quad4_vel) 
+{   
+    if (desired_vel.is_zero()) {
+        return;
+    }
+     // first quadrant: +ve x, +ve y direction
+    if (is_positive(desired_vel.x) && is_positive(desired_vel.y)) {
```

Comment:
These is_positive checks are unnecessary, if it is init to zero any negative number will be lost in the MAX check anyway

Reply:
@IamPete1 yes you are right!  Nice observation.. But I have left it as it is so that its more readable to a new user. It doesn't really add too much cost to our code, and makes everything uniform.

Reply:
I don't think it makes it more readable, it makes 4 lines in to 20, better to add a nice comment explaining what is going on